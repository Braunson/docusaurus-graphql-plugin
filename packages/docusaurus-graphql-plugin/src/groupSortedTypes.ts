import {
  GraphQLNamedType,
  isEnumType,
  isInputObjectType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import { GroupedTypes, InterfaceItem, TypeItem } from "./types";

function groupTypes(types: GraphQLNamedType[]): GroupedTypes {
  return types.reduce<GroupedTypes>(
    (acc, type) => {
      if (type.name.startsWith("__")) {
        return acc;
      }

      if (isObjectType(type)) {
        if (type.name === "Query") {
          return {
            ...acc,
            queries: Object.values(type.getFields()).map((query) => ({
              type: query,
            })),
          };
        }

        if (type.name === "Mutation") {
          return {
            ...acc,
            mutations: Object.values(type.getFields()).map((mutation) => ({
              type: mutation,
            })),
          };
        }

        type.getInterfaces().forEach((inter) => {
          acc.interfaces.forEach((otherInter) => {
            if (otherInter.type.name === inter.name) {
              otherInter.implementedBy.push(type);
            }
          });
        });

        return {
          ...acc,
          objects: acc.objects.concat([{ type }]),
        };
      }

      if (isInterfaceType(type)) {
        const interfaceItem: InterfaceItem = {
          type,
          implementedBy: [],
        };

        acc.objects.forEach(({ type: object }) => {
          object.getInterfaces().forEach((otherInterface) => {
            if (otherInterface.name === interfaceItem.type.name) {
              interfaceItem.implementedBy.push(object);
            }
          });
        });

        return {
          ...acc,
          interfaces: acc.interfaces.concat([interfaceItem]),
        };
      }

      if (isEnumType(type)) {
        return {
          ...acc,
          enums: acc.enums.concat([{ type }]),
        };
      }

      if (isUnionType(type)) {
        return {
          ...acc,
          unions: acc.unions.concat([{ type }]),
        };
      }

      if (isInputObjectType(type)) {
        return {
          ...acc,
          inputObjects: acc.inputObjects.concat([{ type }]),
        };
      }

      if (isScalarType(type)) {
        return {
          ...acc,
          scalars: acc.scalars.concat([{ type }]),
        };
      }

      return acc;
    },
    {
      queries: [],
      mutations: [],
      objects: [],
      interfaces: [],
      enums: [],
      unions: [],
      inputObjects: [],
      scalars: [],
    }
  );
}

function sortByName<T extends TypeItem<{ name: string }>>(types: T[]): T[] {
  return types.sort((a, b) => a.type.name.localeCompare(b.type.name));
}

function sortGroupedTypes(groupedTypes: GroupedTypes): GroupedTypes {
  return {
    queries: sortByName(groupedTypes.queries),
    mutations: sortByName(groupedTypes.mutations),
    objects: sortByName(groupedTypes.objects),
    interfaces: sortByName(groupedTypes.interfaces),
    enums: sortByName(groupedTypes.enums),
    unions: sortByName(groupedTypes.unions),
    inputObjects: sortByName(groupedTypes.inputObjects),
    scalars: sortByName(groupedTypes.scalars),
  };
}

export function groupSortedTypes(types: GraphQLNamedType[]): GroupedTypes {
  return sortGroupedTypes(groupTypes(types));
}
