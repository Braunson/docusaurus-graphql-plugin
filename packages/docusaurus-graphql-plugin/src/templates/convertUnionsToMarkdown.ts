import { MarkdownOptions, UnionItem } from "../types";

export function convertUnionsToMarkdown(
  unions: UnionItem[],
  { getTypePath }: MarkdownOptions
): string {
  const lines: string[] = [];

  unions.forEach(({ type: union }) => {
    lines.push(`## ${union.name}`, `\n\n`);
    lines.push(union.description || "", `\n\n`);

    lines.push(
      `<p style={{ marginBottom: "0.4em" }}><strong>Possible types</strong></p>`,
      `\n\n`
    );
    union.getTypes().forEach((type) => {
      lines.push(`- [${type.name}](${getTypePath(type)})`, `\n`);
    });
    lines.push(`\n`);
  });

  return lines.join("");
}
