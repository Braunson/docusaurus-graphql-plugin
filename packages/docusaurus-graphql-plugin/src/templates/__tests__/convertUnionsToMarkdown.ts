import { buildSchema } from "graphql";
import { groupSortedTypes } from "../../groupSortedTypes";
import { convertUnionsToMarkdown } from "../convertUnionsToMarkdown";

describe("convertUnionsToMarkdown", () => {
  function getTypePath() {
    return "#";
  }

  it("should render unions", () => {
    const schema = buildSchema(`
      type User {
        name: String!
      }
      type Droid {
        name: String!
      }

      """
      Combination of the possible humanoids.
      """
      union Humanoid = User | Droid
    `);
    const { unions } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertUnionsToMarkdown(unions, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });
});
