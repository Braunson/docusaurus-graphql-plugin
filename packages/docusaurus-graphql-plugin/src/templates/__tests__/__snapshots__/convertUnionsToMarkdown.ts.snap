// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertUnionsToMarkdown should render unions 1`] = `
"## Humanoid

Combination of the possible humanoids.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Possible types</strong></p>

- [User](#)
- [Droid](#)

"
`;
